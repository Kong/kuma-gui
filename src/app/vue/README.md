---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

