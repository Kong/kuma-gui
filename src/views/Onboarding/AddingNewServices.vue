<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading title="Add services" />
    </template>
    <template #content>
      <div class="h-full w-full flex justify-evenly items-center">
        <Box
          :active="mode === 'demo'"
          class="cursor-pointer"
          @clicked="update('demo')"
        >
          <div>
            <img src="@/assets/images/new-service-demo.svg?external">
            <div class="ml-3">
              <p class="uppercase font-bold tracking-wider">
                Demo app
              </p>
              <p>Counter application</p>
            </div>
          </div>
        </Box>
        <Box
          :active="mode === 'manually'"
          class="cursor-pointer"
          @clicked="update('manually')"
        >
          <div class="cursor-pointer">
            <img src="@/assets/images/new-service-manually.svg?external">
            <div class="ml-3">
              <p class="uppercase font-bold tracking-wider">
                Manually
              </p>
              <p>After this wizzard</p>
            </div>
          </div>
        </Box>
      </div>
    </template>
    <template #navigation>
      <OnboardingNavigation
        :next-step="nextStep"
        previous-step="onboarding-populating-mesh"
      />
    </template>
  </OnboardingPage>
</template>

<script>
import { mapGetters, mapMutations } from 'vuex'
import OnboardingNavigation from '@/views/Onboarding/components/OnboardingNavigation'
import OnboardingHeading from '@/views/Onboarding/components/OnboardingHeading'
import OnboardingPage from '@/views/Onboarding/components/OnboardingPage'
import Box from '@/views/Onboarding/components/Box'

export default {
  name: 'AddingNewServices',
  components: {
    OnboardingNavigation,
    OnboardingHeading,
    OnboardingPage,
    Box,
  },
  metaInfo() {
    return {
      title: 'Adding new services',
    }
  },
  computed: {
    ...mapGetters({
      onboardingMode: 'onboarding/getMode',
    }),

    nextStep() {
      if (this.mode === 'manually') {
        return 'onboarding-completed'
      }

      return 'onboarding-adding-services-code'
    },
    mode: {
      get() {
        return this.onboardingMode
      },
      set(value) {
        this.update(value)
      },
    },
  },
  methods: {
    ...mapMutations({
      update: 'onboarding/UPDATE_MODE',
    }),
  },
}
</script>
