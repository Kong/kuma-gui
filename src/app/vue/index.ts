import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
