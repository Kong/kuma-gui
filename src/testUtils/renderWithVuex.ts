import { render, RenderOptions, RenderResult } from '@testing-library/vue'
import merge from 'lodash/merge'
import { StoreOptions } from 'vuex'
import Vue, { ComponentOptions } from 'vue'
import Store from '@/store'

function renderWithVuex(
  Component: typeof Vue | ComponentOptions<Vue>,
  options: RenderOptions<Vue> = { store: {}, routes: [] },
): RenderResult {
  const store: StoreOptions<any> = Store()

  // Render the component and merge the original store and the custom one
  // provided as a parameter. This way, we can alter some behaviors of the
  // initial implementation.

  return render(Component, { ...options, store: merge(store, options.store), routes: options.routes })
}

export default renderWithVuex
