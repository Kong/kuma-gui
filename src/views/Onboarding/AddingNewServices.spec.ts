import { screen } from '@testing-library/vue'
import userEvent from '@testing-library/user-event'
import renderWithVuex from '@/testUtils/renderWithVuex'
import AddingNewServices from './AddingNewServices.vue'

describe('AddingNewServices.vue', () => {
  it('renders snapshot', () => {
    const { container } = renderWithVuex(AddingNewServices)

    expect(container).toMatchSnapshot()
  })

  it('changes selected box', async () => {
    renderWithVuex(AddingNewServices)

    const demoRadio = screen.getAllByTestId('box')[0]
    const manuallyRadio = screen.getAllByTestId('box')[1]

    expect(demoRadio.className).toMatch('box--active')
    expect(manuallyRadio.className).not.toMatch('box--active')

    await userEvent.click(screen.getByText(/After this wizzard/))

    expect(demoRadio.className).not.toMatch('box--active')

    expect(manuallyRadio.className).toMatch('box--active')
  })
})
