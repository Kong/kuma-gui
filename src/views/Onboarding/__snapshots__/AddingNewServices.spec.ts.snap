// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddingNewServices.vue renders snapshot 1`] = `
<div>
  <div>
    <div
      class="onboarding-container"
    >
      <div
        class="onboarding-container__header"
      >
        <div
          class="relative"
        >
          <h1
            class="onboarding-title"
          >
            
    Add services
  
          </h1>
           
          <!---->
        </div>
      </div>
       
      <div
        class="onboarding-container__content"
      >
        <div
          class="w-full"
        >
          <div
            class="h-full w-full flex justify-evenly items-center"
          >
            <div
              class="cursor-pointer box box--active"
              data-testid="box"
            >
              <div>
                <img
                  src="@/assets/images/new-service-demo.svg?external"
                />
                 
                <div
                  class="ml-3"
                >
                  <p
                    class="uppercase font-bold tracking-wider"
                  >
                    
              Demo app
            
                  </p>
                   
                  <p>
                    Counter application
                  </p>
                </div>
              </div>
            </div>
             
            <div
              class="cursor-pointer box"
              data-testid="box"
            >
              <div
                class="cursor-pointer"
              >
                <img
                  src="@/assets/images/new-service-manually.svg?external"
                />
                 
                <div
                  class="ml-3"
                >
                  <p
                    class="uppercase font-bold tracking-wider"
                  >
                    
              Manually
            
                  </p>
                   
                  <p>
                    After this wizzard
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
       
      <div
        class="mt-4 flex items-center flex-col sm:flex-row justify-between"
      >
        <a
          aria-current="page"
          class="k-button navigation-button navigation-button--back router-link-exact-active router-link-active primary"
          data-v-38fa8ecf=""
          href="#/"
          type="button"
        >
          
    Back
  
        </a>
         
        <div>
          <button
            class="k-button skip-button small btn-link"
            data-v-38fa8ecf=""
            type="button"
          >
            
      Skip Setup
    
          </button>
           
          <a
            aria-current="page"
            class="k-button navigation-button navigation-button--next router-link-exact-active router-link-active primary"
            data-v-38fa8ecf=""
            href="#/"
            type="button"
          >
            
      Next
    
          </a>
        </div>
      </div>
    </div>
     
    <div
      class="background-image"
    />
  </div>
</div>
`;
